packages/cli/src/util/dev/server.ts

```diff
@@ -675,31 +675,48 @@ export default class DevServer {
       this.getLocalEnv('.env.build', vercelConfig.build?.env),
     ]);
 
-    let allEnv = { ...buildEnv, ...runEnv };
-
     // If no .env/.build.env is present, use cloud environment variables
-    if (Object.keys(allEnv).length === 0) {
+    if (Object.keys({ ...runEnv, ...buildEnv }).length === 0) {
       const envValues = { ...this.envValues };
-      if (this.address.host) {
-        envValues['VERCEL_URL'] = this.address.host;
-      }
-      allEnv = { ...envValues };
       runEnv = { ...envValues };
       buildEnv = { ...envValues };
     }
 
     // legacy NOW_REGION env variable
     runEnv['NOW_REGION'] = 'dev1';
     buildEnv['NOW_REGION'] = 'dev1';
-    allEnv['NOW_REGION'] = 'dev1';
 
-    // mirror how VERCEL_REGION is injected in prod/preview
-    // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
-    // and should not contain VERCEL_REGION
     if (this.projectSettings?.autoExposeSystemEnvs) {
+      // mirror how VERCEL_REGION is injected in prod/preview
+      // only inject in `runEnvs`, because `allEnvs` is exposed to dev command
+      // and should not contain VERCEL_REGION
       runEnv['VERCEL_REGION'] = 'dev1';
+
+      // expose VERCEL_ENV and VERCEL_URL to dev command
+      runEnv['VERCEL_ENV'] = 'development';
+      buildEnv['VERCEL_ENV'] = 'development';
+      if (this.address.host) {
+        runEnv['VERCEL_URL'] = this.address.host;
+        buildEnv['VERCEL_URL'] = this.address.host;
+      }
+
+      // automaticaly expose VERCEL_URL, VERCEL_ENV
+      // as prefixed (framework) env vars
+      if (this.projectSettings?.framework) {
+        const frameworkSlug = this.projectSettings.framework;
+        const framework = frameworkList.find(f => f.slug === frameworkSlug);
+        const envPrefix = framework?.envPrefix;
+
+        if (envPrefix) {
+          buildEnv = {
+            ...buildEnv,
+            ...getPrefixedEnvVars({ envPrefix: envPrefix, envs: buildEnv }),
+          };
```

Comment:
```suggestion
           const prefixedEnvs = getPrefixedEnvVars({
             envPrefix,
             envs: buildEnv
           });

           for (const [key, value] of Object.entries(prefixedEnvs)) {
             buildEnv[key] = value;
           }
```

Lets avoid iterating over buildEnv when we can mutate instead.

This also follows the same pattern for usage of `getPrefixedEnvVars()` in other places.